### PR TITLE
Fix IntegerOverflow exception in postings encoding as group-varint

### LINKED ISSUE #13373: DataOutput.writeGroupVInts throws IntegerOverflow exception during merging
### Description

As being discussed on email list that `DataOutput.writeGroupVInts` throws as IntegerOverflow exception.  The goal is to find out the main reason and also to improve the exception message.

```
Exception in thread "Lucene Merge Thread #202"
org.apache.lucene.index.MergePolicy$MergeException:
java.lang.ArithmeticException: integer overflow at
org.apache.lucene.index.ConcurrentMergeScheduler.handleMergeException(ConcurrentMergeScheduler.java:735) at
org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:727)
Caused by: java.lang.ArithmeticException: integer overflow at 
java.base/java.lang.Math.toIntExact(Math.java:1135) at 
org.apache.lucene.store.DataOutput.writeGroupVInts(DataOutput.java:354) at 
org.apache.lucene.codecs.lucene99.Lucene99PostingsWriter.finishTerm(Lucene99PostingsWriter.java:379) at 
org.apache.lucene.codecs.PushPostingsWriterBase.writeTerm(PushPostingsWriterBase.java:173) at 
org.apache.lucene.codecs.lucene90.blocktree.Lucene90BlockTreeTermsWriter$TermsWriter.write(Lucene90BlockTreeTermsWriter.java:1097) at
org.apache.lucene.codecs.lucene90.blocktree.Lucene90BlockTreeTermsWriter.write(Lucene90BlockTreeTermsWriter.java:398) at 
org.apache.lucene.codecs.FieldsConsumer.merge(FieldsConsumer.java:95) at
org.apache.lucene.codecs.perfield.PerFieldPostingsFormat$FieldsWriter.merge(PerFieldPostingsFormat.java:205) at org.apache.lucene.index.SegmentMerger.mergeTerms(SegmentMerger.java:209) at
org.apache.lucene.index.SegmentMerger.mergeWithLogging(SegmentMerger.java:298) at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:137) at
org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:5252) at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:4740) at
org.apache.lucene.index.IndexWriter$IndexWriterMergeSource.merge(IndexWriter.java:6541) at
org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:639) at
org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:700)
```

More context from the reporter

> Looking deeper into this. I think we overflowed a term frequency field.
> Looking in some statistics, in a previous release we had 1,288,526,281
> of a certain field, this would be larger now. Each of these would have
> had a limited set of values. But crucially nearly all of them would have
> had the term "positional" or "non-positional" added to the document.
> 
> There is no good reason to do this today, we should just turn this into
> a boolean field and update the UI. I will do this and report back.
> 
> Do you think that a patch for a try/catch for a more informative log
> message be appreciated by the community? e.g. mentioning the field name
> in the exception?

> The index that had an issue when merging into one segment definitely had
> more than 1 billion times the word "positional" in it. I hope to be able
> to give a closer number once re-indexing finished with a "work-around".
> 
> Of course the "work-around" is to just fix this correctly by not having
> that word so often in the index and definitely not as docs, freqs and
> postings.
> 
> For background information.
> 
> The use case was to find a set of documents that where either
> "positional" or "non-positional". This was present in the first check in
> of our code 18 years ago! since then our data has grown a bit ;) The
> code was using Lucene 1.4.3 at that time. Users would search using this
> as what now would be a facet `type:positional`. I changed this to a
> field only IndexOptions.DOCS which is called 'positional' and searched
> as `positional:yes` rewriting the previous query syntax behind the scene
> to not break any user tools.

### Version and environment details

_No response_

### PR DESCRIPTION
**Pull request by easyice**
Closes: https://github.com/apache/lucene/issues/13373

This exception occurs because a negative  integer value stores as  positive long.  In line 376, after a long value `<< 1`, if the sign bit of the **integer** value is 1, it will be a  negative number as integer, but a positive numbers as long, when we stores this value as positive long, it would cause `Math.toIntExact` to throw `ArithmeticException` exception.

https://github.com/apache/lucene/blob/f12e4899bf0420693e4f524a515dafcf0f21a3d3/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsWriter.java#L373-L379

POC code:

```java
    public void testDelta() {
        int delta = 1 << 30;
        long deltaL = delta;
        deltaL = (deltaL << 1) | 1;
        System.out.println(deltaL > Integer.MAX_VALUE); // true
        System.out.println(deltaL); // 2147483649
        System.out.println((int) deltaL); // -2147483647
        Math.toIntExact(deltaL); // ArithmeticException: integer overflow
    }
```

TODO:

- [x] add more tests

### CODE REVIEW

**Comment by easyice** [comment]
The essence of this issue is how to deal with the integer value with the sign bit as 1 (like this integer overflow case). We have two options. 
* The first is pass a negative long value to `writeGroupVInts`, and also return a negative value in `readGroupVInt`. Previously, we didn't consider encoding a negative value in group-varint. Currently, `readGroupVInt` may return a positive or negative value in different implementations. 
* The second approach is only pass positive numbers to `writeGroupVInts` / `readGroupVInt`, and replace `Math.toIntExact` with `(int)`.

The first approach feels more reasonable.

**Comment by jpountz** [comment]
Thanks for looking into it! Your approach works, but I'm tempted to fix it the other way around, by no longer checking if values are in the expected range with `Math.toIntExact` but rather by checking that the value is in the `[0, 2**32)` range?

**Comment by easyice** [comment]
That's also a good idea! by this approach we can make `writeGroupVInts `/`readGroupVInt` use positive only. it's actually handled as an unsigned integer, so we don't need to consider the case of encoding negative in group-varint.  i will update the code.

File: lucene/core/src/java/org/apache/lucene/util/GroupVIntUtil.java [hunk 1]
```diff
@@ -118,6 +120,13 @@ private static int numBytes(int v) {
     return Integer.BYTES - (Integer.numberOfLeadingZeros(v | 1) >> 3);
   }
 
+  private static int toInt(long value) {
+    if (value < 0 || value > 0xFFFFFFFFL) {
+      throw new ArithmeticException("integer overflow");
+    }
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/GroupVIntUtil.java:126]: Maybe use `Long.compareUnsigned`? (`if (Long.compareUnsigned(value, 0xFFFFFFFFL) > 0)`)

**Review by jpountz — APPROVED** [review]

File: lucene/test-framework/src/java/org/apache/lucene/tests/store/BaseDirectoryTestCase.java [hunk 2]
```diff
@@ -1442,6 +1442,19 @@ public void testListAllIsSorted() throws IOException {
     }
   }
 
+  public void testGroupVIntOverflow() throws IOException {
+    try (Directory dir = getDirectory(createTempDir("testGroupVIntOverflow"))) {
+      final int v = 1 << 30;
+      final long[] values = new long[4];
+      values[0] = v;
+      values[0] <<= 1; // values[0] = 2147483648 as long, but as int it is -2147483648
```
[the remark this answers is not in the document]
Comment (easyice) [reply lucene/test-framework/src/java/org/apache/lucene/tests/store/BaseDirectoryTestCase.java]: fixed

File: lucene/core/src/java/org/apache/lucene/util/GroupVIntUtil.java (hunk 1 above)
Comment (easyice) [reply lucene/core/src/java/org/apache/lucene/util/GroupVIntUtil.java:126]: Great!

**Review by easyice — COMMENTED** [review]
Thanks for reviewing!

File: lucene/core/src/java/org/apache/lucene/store/DataOutput.java [hunk 3]
```diff
@@ -328,9 +328,12 @@ public void writeSetOfStrings(Set<String> set) throws IOException {
...
    * Encode integers using group-varint. It uses {@link DataOutput#writeVInt VInt} to encode tail
    * values that are not enough for a group. we need a long[] because this is what postings are
-   * using, all longs are actually required to be integers.
+   * using, all longs are actually required to be integers. Negative numbers are supported, but
+   * should be avoided.
    *
-   * @param values the values to write
+   * @param values the values to write. Note: if original integer is negative, it should also be
+   *     negative as long, not positive which is greater than Integer.MAX_VALUE, that will cause
+   *     integer overflow exception in {@link Math#toIntExact(long)}.
```
[the remark this answers is not in the document]
Comment (easyice) [reply lucene/core/src/java/org/apache/lucene/store/DataOutput.java]: This change has reverted, there is no change in `DataOutput` in current fix approach.

**Review by jpountz — APPROVED** [review]

**Comment by mikemccand** [comment]
It's hard for me to tell what the expected user impact here is?  Does the exception happen because the remainder part of a postings list (after all length 128 blocks are done), which we now encode with GroupVInt, had a docID delta that was `>= 1<<30`, when the postings are also storing freqs?

I guess because we don't see too many users reporting this, it is likely rare-ish.  But is the GroupVInt change released in 9.x?  Is this maybe enough to warrant a bugfix release if so?

**Comment by easyice** [comment]
> Does the exception happen because the remainder part of a postings list (after all length 128 blocks are done), which we now encode with GroupVInt, had a docID delta that was >= 1<<30, when the postings are also storing freqs?

Yes, exactly. I guess the docID delta that was `>= 1<<30` s likely rare-ish

> But is the GroupVInt change released in 9.x? Is this maybe enough to warrant a bugfix release if so?

GroupVInt was released at 9.9.0 https://lucene.apache.org/core/9_9_0/changes/Changes.html#v9.9.0.optimizations
+1 for bugfix release, what do you think?

**Comment by jpountz** [comment]
+1 to a bugfix release

**Comment by jpountz** [comment]
Can you backport to the 9.10 branch?

**Comment by easyice** [comment]
Okay, I will backport to 9.10/branch_9x.

**Pull request merged**